File: BAEKJOON/all_problems/2K/2469.cpp
#include <bits/stdc++.h>
using namespace std;

void move_by_ladder(int k, string &s, string ladder){
    for(int i=0; i<(k-1); ++i){
        if(ladder[i] == '-'){
            swap(s[i], s[i+1]);
        }
    }
}

void solve(){
    int k; cin >> k;
    int n; cin >> n;
    string ts; cin >> ts;
    string ladders[n];
    for(auto &ladder:ladders) cin >> ladder;
    // for(auto &ladder:ladders) cout << ladder << '\n';

    int question_mark_index = -1;
    for(int i=0; i<n; ++i){
        if(ladders[i][0] == '?')
            question_mark_index = i;
    }
    // cout << question_mark_index << '\n';

    string bs="";
    for(int i=0; i<k; ++i){
        // bs += "A"+i;       //
        bs += 'A'+i;
    }
    // cout << bs << '\n';

    for(int i=0; i<question_mark_index; ++i){
        move_by_ladder(k, bs, ladders[i]);
    }
    // cout << bs << '\n';

    for(int i=n-1; i>=(question_mark_index+1); --i){
        move_by_ladder(k, ts, ladders[i]);
    }
    // cout << ts << '\n';

    int i = 0;
    string ans = "";
    bool pos = true;

    while(i < (k-1)){
        if(bs[i] == ts[i]){
            ans += '*';
            ++i;
        }else if(bs[i] == ts[i+1] and bs[i+1] == ts[i]){
            ans += '-';
            if(i+1 < (k-1)) ans += '*';
            i += 2;
        }else{
            pos = false;
            break;
        }
    }

    if(!pos){
        for(int i=0; i<(k-1); ++i) cout << 'x';
        return;
    }
    cout << ans;
}

int main(void){
    ios::sync_with_stdio(0);
    cin.tie(0);
    solve();
    return 0;
}

File: BAEKJOON/all_problems/2K/2469.py
import sys
input = sys.stdin.readline

def move_by_ladder(k, s, ladder):
    for i in range(k-1):
        c = ladder[i]
        if(c == '-'):
            s[i], s[i+1] = s[i+1], s[i]
    return s

def solve():
    k = int(input())
    n = int(input())
    target_s = list(input().rstrip())
    # print(target_s)

    base_s = [chr(65+i) for i in range(k)]
    # print("base_s", base_s)

    ladders = [input().rstrip() for _ in range(n)]
    question_mark_index = 0
    for i in range(n):
        if(ladders[i][0] == '?'):
            question_mark_index = i
    # print(question_mark_index)

    for i in range(question_mark_index):
        ladder = ladders[i]
        base_s = move_by_ladder(k, base_s, ladder)
    # print(base_s)

    for i in range(question_mark_index+1, n):
        ladder = ladders[n-1+question_mark_index+1-i]
        target_s = move_by_ladder(k, target_s, ladder)
    # print("target_s", target_s)

    ans = ['_' for _ in range(k-1)]
    # print("ans", ans)

    i = 0
    while(i < (k-1)):
        if(base_s[i] == target_s[i]):
            ans[i] = '*'
            i += 1
        elif(base_s[i] == target_s[i+1] and\
            base_s[i+1] == target_s[i]):
            if(ans[i] != '-'):
                ans[i] = '-'
                if(i+1 < (k-1)):
                    ans[i+1] = '*'
                i += 2
            else:
                for j in range(k-1):
                    ans[j] = 'x'
                break
        else:
            for j in range(k-1):
                ans[j] = 'x'
            break
    print(''.join(ans))

solve()